File: shells/tvos/TinyTrackParty/Render/PerfOverlay.swift
import SwiftUI

/// The frame-cost readout, and the tvOS half of `render/PerfHud.js`.
///
/// **IT MEASURES; IT DOES NOT JUDGE.** The ring, its trim, the warm-up filter,
/// the two rates, the percentiles, the drop and skip counts and the health
/// verdict are all `ttp_perf.h` (`libttp-runtime/ttp/perf_stats.h` carries the
/// reasoning, and the `perf` ctest executes it on every leg). What is left here
/// is what only this platform can do: read `CADisplayLink`'s clock and the
/// renderer's profile buffer, hand them over, and draw the answer. Three
/// hand-written folds used to sit in three shells, each with a comment claiming
/// they agreed, and by the time they were replaced this one folded skipped
/// presents into its verdict and the other two did not — so the same run was
/// amber on a television and green in a browser.
///
/// THE GPU NUMBER IS REAL HERE NOW, and it took a Filament fork commit to make
/// it so. `MetalTimerQueryFence` measured the interval between two shared-event
/// fence signals with `clock::now()` — CPU wall time between two points in the
/// submission stream, which under vsync tracks the PRESENT CADENCE and not the
/// work. It reported about 16.7 ms for every frame a 60 Hz panel delivered,
/// idle or saturated, so feeding it in would have been worse than absent: a
/// number that is always exactly the budget is a number that always says
/// "late". The pinned fork takes GPUStartTime/GPUEndTime off the command
/// buffers instead (native/filament.pin; upstream PR google/filament#10338).
///
/// IT IS NOT ONLY A READOUT. The render scale folds off this same monitor, so
/// what is passed here steers the television's resolution — which is why this
/// arrived WITH `render_scale.h`'s present-record veto and not before it. On
/// this device the GPU p95 at 4 players sits past the down threshold while the
/// box presents 60/60 with zero skips, and without that veto, wiring the timer
/// in cost a whole rung of picture.
///
/// Absent (<= 0) is still the honest report for a platform with no timer, and
/// it is not a free choice: with no GPU stat the verdict's overshoot term falls
/// back to the present interval's p95, so a plausible substitute would silently
/// move the bar.
///
/// OFF UNTIL ASKED FOR, as the web's and Android's are. `-ttpPerf 1` at launch
/// is the switch ([enabledAtLaunch]) and `GameCoordinator.boot()` reads it;
/// everything below is inert while hidden, and the CPU term is not even sampled.
///
/// A LAUNCH ARGUMENT AND NOT A BUTTON, which is this platform's constraint and
/// not a preference: every button the Siri Remote can reach is already spoken
/// for (Menu walks back, Play/Pause is the pause, select and the d-pad belong to
/// the focus chain), so a toggle could only be built by taking one off a player.
/// It is the same knob shape as `-ttpRenderScale` and `-ttpFeatures`
/// (`DisplayHost`), it survives a force-stop the way a keypress would not, and
/// the bench turns the panel on for itself ([bench]) without being asked.
@MainActor
final class PerfMonitor: ObservableObject {

    /// Published at 4 Hz, not 60: an `@Published` write per frame would redraw
    /// the overlay as often as the scene it is measuring.
    private static let textInterval = 0.25
    /// The bench contract's cadence — one `TtpPerf <json>` line a second, the
    /// same on all three shells so one parser reads all three logs.
    private static let logInterval = 1.0

    /// `-ttpPerf 1`: put the readout up for this launch. Absent means off, which
    /// is what a player gets — a television is not a browser with a console, and
    /// a diagnostic block over the corner of the picture is the whole of what
    /// they would see of it.
    ///
    /// LIVE IN RELEASE ON PURPOSE, exactly as Android's `debug.ttp.*` knobs are:
    /// the build that ships is the only one with the real optimisation level, and
    /// an instrument compiled out of it cannot measure the thing people run.
    static let enabledAtLaunch = UserDefaults.standard.bool(forKey: "ttpPerf")

    @Published private(set) var visible = false
    @Published private(set) var lines: [String] = []
    @Published private(set) var tint = Color.green

    private var lastText: Double = 0
    private var lastLog: Double = 0
    private var benching = false
    /// What is being driven, for a sweep that spans a catalogue. Only a bench
    /// knows it (the driver names the circuit); a party run reports null.
    private var track: String?
    private var totalIndex: Int?
    private var namesResolved = false

    /// Switch the readout on.
    func show() {
        visible = true
        lastText = 0
        reset()
    }

    /// Drop the shared window — stale history is worse than none. Called
    /// whenever what is being measured changes underneath: the readout coming
    /// on, a resize and a scene build (`DisplayHost.applyResize` and `sceneBuilt`).
    func reset() { ttp_perf_reset() }

    /// Start a BENCHED run on `track`: the same readout, additionally logged
    /// once a second as `TtpPerf <json>` — one JSON object per line, the very
    /// bytes the panel above is drawn from, so a screenshot and a logged number
    /// cannot disagree. `Log.i("TtpPerf", …)` on Android and `console.log` on
    /// the web emit the identical shape; `scripts/perf-race.tvos.mjs` reads this
    /// one back off `devicectl --console`.
    func bench(track: String) {
        benching = true
        self.track = track.isEmpty ? nil : track
        lastLog = 0
        show()
    }

    /// One display-link TICK, drawn or not. `interval` is the elapsed time since
    /// the previous tick (not the link's nominal cadence), because a missed
    /// vsync is exactly what this is here to show; `presented` is whether that
    /// tick put a new picture on the panel.
    func record(now: Double, interval: Double, presented: Bool,
                cells: Int, pixels: CGSize, dpr: CGFloat) {
        // FED WHETHER OR NOT ANYONE IS WATCHING. The render scale folds off this
        // same monitor (`ttp_perf.h`), so a window kept only while the panel is
        // up would leave the rule deciding a television's resolution off an
        // empty ring the moment someone pressed the toggle. The cost of the
        // sample is a push into a 120-frame ring; the cost of this object is
        // `publish`, which is behind the interval below.
        //
        // SECONDS on this side, MILLISECONDS on that one: `CADisplayLink`
        // timestamps are seconds and `ttp_perf_sample` takes ms.
        //
        // The cpu sample is dropped (0, i.e. absent) on a skip rather than
        // repeated: the renderer returns before writing its `total`, so the
        // profile still holds the last DRAWN frame, and folding it in again
        // would weight the median hardest under exactly the load that causes
        // skips. It is also absent while nothing is drawing the readout, since
        // it feeds no other reader.
        //
        // THE GPU SAMPLE IS FED WHENEVER A FRAME DREW, watched or not, because
        // unlike the cpu one it has a second reader: the render scale. Gating it
        // on the panel being up would hand the rule a different signal depending
        // on whether anybody was looking at it.
        ttp_perf_sample(now * 1000, interval * 1000, presented ? 1 : 0,
                        presented && visible ? (cpuTotalMs() ?? 0) : 0,
                        presented ? ttp_display_gpu_ms() : 0)
        guard visible, now - lastText >= Self.textInterval else { return }
        lastText = now
        publish(now: now, cells: cells, pixels: pixels, dpr: dpr)
    }

    /// Last frame's CPU total, out of the renderer's own strided profile array.
    /// The section names are fixed for the life of the process, so the index is
    /// resolved once rather than re-split per frame.
    private func cpuTotalMs() -> Double? {
        if !namesResolved {
            namesResolved = true
            totalIndex = TTP.strOrEmpty(ttp_display_profile_names())
                .split(separator: Character(",")).map(String.init).firstIndex(of: "total")
        }
        guard let i = totalIndex, let profile = ttp_display_profile() else { return nil }
        return profile[i]
    }

    /// ONE readout, drawn and logged. The buffer size and the CELL COUNT ride it
    /// because GPU cost scales with cells and pixels together: a logged number
    /// carrying neither is not comparable to any other logged number. They are
    /// not in the drawn rows — a human looking at the television can already see
    /// how many cells are on it.
    private func publish(now: Double, cells: Int, pixels: CGSize, dpr: CGFloat) {
        let json = TTP.strOrEmpty(ttp_perf_readout_json(
            Int32(cells), Int32(pixels.width), Int32(pixels.height), Double(dpr), track))
        if benching && now - lastLog >= Self.logInterval {
            lastLog = now
            print("TtpPerf \(json)")
        }
        let r = TTP.obj(json)
        let w = num(r["width"]), h = num(r["height"])
        // WARMING IS NOT A FRAME RATE (perf_stats.h): the first presents of a run
        // are shader compiles and first uploads, and the fold discards them
        // rather than reporting them. Saying so beats drawing 0/0 fps, which a
        // viewer reads as a stall in the one second where it never is.
        guard !(r["warming"] as? Bool ?? false) else {
            lines = ["\(w)×\(h)", "warming up"]
            tint = Self.good
            return
        }
        let fps = num(r["fps"]), hz = num(r["hz"])
        let skips = num(r["skips"]), drops = num(r["drops"])
        // Cost as a SHARE OF BUDGET USED, low is good, the same sense as the
        // web's readout.
        let budget = r["budgetMs"] as? Double ?? 0
        let share = { (stat: String) -> Double? in
            guard budget > 0, let s = (r[stat] as? [String: Any])?["p50"] as? Double,
                  s > 0 else { return nil }
            return s
        }
        let pct = { (ms: Double?) in ms.map { "\(Int((100 * $0 / budget).rounded()))%" } ?? "n/a" }
        let cpuText = pct(share("cpu"))
        // THE GPU COST CARRIES ITS MILLISECONDS TOO, which the web's readout does
        // not. A browser has devtools a keypress away; a television has this
        // panel and nothing else, and the share alone hides the thing this
        // platform most needs to see — the A10X downclocks whenever a frame
        // leaves idle, so the same picture prices differently at 60 fps than it
        // does under load, and a percentage of a budget that also moved with the
        // present rate cannot show that. Absent on a platform with no timer,
        // which is what "n/a" means here and not "free".
        let gpuMs = share("gpu")
        let gpuText = gpuMs.map { "\(pct($0)) (\(String(format: "%.1f", $0)) ms)" } ?? "n/a"
        // TWO ROWS, the two the web settled on: the surface and the cadence on
        // one line, the cost on the other. `fps/hz` rather than one rate because
        // the PAIR is the diagnosis — 60/60 is a healthy panel, 41/60 is a GPU
        // that cannot hold this resolution, and one number cannot tell them
        // apart at all.
        lines = [
            "\(w)×\(h) · \(fps)/\(hz) fps · "
                + "\(skips) skip\(skips == 1 ? "" : "s") · \(drops) drop\(drops == 1 ? "" : "s")",
            // BOTH COSTS AS A SHARE OF BUDGET USED, low is good, and in the
            // web's order (gpu first): they do NOT sum — the CPU builds frame
            // N's commands while the GPU still draws N-1 — so the larger of the
            // two is the one to cut, and reading them off two shells has to be
            // the same motion.
            "gpu \(gpuText) · cpu \(cpuText)"
        ]
        // DIAGNOSTIC colours, not chrome: the sticker palette's veto on amber
        // does not reach a debug overlay, and the three readouts agreeing is
        // worth more here than any of them matching the theme.
        switch r["verdict"] as? String {
        case "bad": tint = Color(red: 1, green: 0.42, blue: 0.42)
        case "warn": tint = Color(red: 1, green: 0.82, blue: 0.4)
        default: tint = Self.good
        }
    }

    private static let good = Color(red: 0.49, green: 0.99, blue: 0.54)

    /// A count out of the readout. `JSONSerialization` hands every number back
    /// as an `NSNumber`, and the readout's counts are whole by construction.
    private func num(_ v: Any?) -> Int { (v as? NSNumber)?.intValue ?? 0 }
}

/// Drop this over the scene (`.overlay(alignment: .bottomTrailing)`); SwiftUI's
/// safe area keeps it inside the TV's overscan. It renders nothing at all while
/// the monitor is hidden.
struct PerfOverlay: View {
    @ObservedObject var monitor: PerfMonitor

    var body: some View {
        if monitor.visible {
            VStack(alignment: .leading, spacing: 2) {
                ForEach(Array(monitor.lines.enumerated()), id: \.offset) { Text($0.element) }
            }
            .font(.system(size: 22, weight: .semibold, design: .monospaced))
            .foregroundColor(monitor.tint)
            .padding(.horizontal, 12)
            .padding(.vertical, 9)
            .background(Color.black.opacity(0.58))
            .clipShape(RoundedRectangle(cornerRadius: 8))
            .allowsHitTesting(false)
        }
    }
}

File: shells/androidtv/app/src/main/kotlin/com/couchgames/tinytrackparty/PerfOverlay.kt
package com.couchgames.tinytrackparty

import android.util.Log
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import java.nio.ByteOrder
import java.util.Locale

/**
 * The frame-cost readout, and the Android half of `render/PerfHud.js` and
 * `Render/PerfOverlay.swift`. Read those two first — the reasoning about what a
 * clock here can and cannot mean is written down once, there.
 *
 * THREE CLOCKS, AND THEY DO NOT MEASURE THE SAME THING.
 *
 *  - The PRESENT interval, which is the cadence the TV actually showed. Under
 *    vsync it is a plateau, so it says nothing about headroom, but it is the only
 *    thing that says whether a budget was MISSED, which is the part a human
 *    feels. It counts PRESENTS, not Choreographer callbacks — see [DisplayHost],
 *    because the difference between those two is a readout saying 60 on a box
 *    showing 15, and this shell shipped that way.
 *  - CPU, from `ttp_display_profile`: the C++ building the frame's input from the
 *    live `Game` and issuing its draws. Under two milliseconds here, always.
 *  - GPU, from `ttp_display_gpu_ms` — the GL backend's own timer query, and the
 *    one number that can see headroom. It is REAL on this platform, unlike the
 *    two siblings, both of which say so in their own headers.
 *
 * The costs do NOT sum: the CPU builds frame N's commands while the GPU is still
 * drawing N-1. Whichever is larger is the one to cut, and on this box it has
 * never once been the CPU.
 *
 * WHAT THIS ONE ADDS OVER ITS TWO SIBLINGS is the per-section split, and the
 * reason is a measurement rather than taste: on the web the total is under a
 * millisecond and the sections below it are quantization noise, while here the
 * frame costs tens of milliseconds and WHICH section holds them is the question.
 *
 * ON IN DEBUG BUILDS, and toggled with KEYCODE_INFO — which this box's remote does
 * not carry, so in practice it is `adb shell input keyevent 165`. Everything is
 * inert while hidden: [record] returns before it touches the profile ABI.
 *
 * MAIN THREAD ONLY, like everything that touches a `ttp_display_*`.
 */
object PerfMonitor {

    /**
     * The bar, and the denominator of every percentage. Declared in
     * [DisplayHost], which also OWNS this value: at every-other-vsync pacing
     * (`debug.ttp.hz 30`) it doubles, or every healthy 33 ms frame would read
     * as a drop and every percentage would lie by 2x.
     */
    var budgetMs = DisplayHost.BUDGET_MS
    /**
     * The cost stats fold at most this many frames, AND at most [WINDOW_MS] of
     * them. The frame cap alone is a trap on a slow box, and it cost a whole
     * ablation sweep: 120 frames at 12 fps is TEN SECONDS of history, so every
     * arm of the sweep read as a blend of itself and the arm before it — which
     * is exactly the situation the readout is for. The time bound is what makes
     * a reading describe the last few seconds at any frame rate.
     */
    private const val WINDOW = 120
    private const val WINDOW_MS = 3000.0
    private const val TEXT_INTERVAL_MS = 250.0
    private const val LOG_INTERVAL_MS = 1000.0

    /**
     * Sections named on the second line, in the order they are shown. A SUBSET of
     * `ttp_display_profile_names()`, picked because the rest are consistently
     * under a tenth of a millisecond here and a row nobody reads costs the same
     * screen space as one they do. `total` is the first line's `cpu`. The spike
     * lines ([logSpikes]) are NOT limited to this list — a phase that is nothing
     * at the median is exactly what they exist to catch.
     */
    private val SHOWN = listOf("cars", "world", "skids", "decalUp", "build",
        "cellRender", "present", "endFrame")

    /** The backend's own GPU milliseconds, over the same window. See `ttp_display_gpu_ms`. */
    private val gpu = ArrayList<Double>(WINDOW)

    var visible by mutableStateOf(false)
        private set
    var lines by mutableStateOf(listOf<String>())
        private set
    var tint by mutableStateOf(Color.Green)
        private set

    /** Stale history is worse than none: a window straddling the toggle describes neither state. */
    fun show() { visible = true; reset() }
    fun hide() { visible = false }
    fun toggle() { if (visible) hide() else show() }

    private fun reset() {
        stamps.clear(); intervals.clear(); sections.clear(); gpu.clear()
        lastText = 0.0; lastLog = 0.0
        ticks = 0; skipped = 0
    }

    /** Callbacks and, of those, the ones Filament declined to draw. See [record]. */
    private var ticks = 0
    private var skipped = 0

    /** Sample timestamps, parallel to [intervals] — the time bound reads these. */
    private val stamps = ArrayList<Double>(WINDOW)
    private val intervals = ArrayList<Double>(WINDOW)
    /** Per shown section, the trailing window of per-frame milliseconds. */
    private val sections = HashMap<String, ArrayList<Double>>()
    private var lastText = 0.0
    private var lastLog = 0.0

    /** Resolved once: the section names are fixed for the life of the process. */
    private var names: List<String>? = null

    /**
     * One Choreographer callback. `presented` is whether it actually reached the
     * screen — see the callback in [DisplayHost] for why the two are not the same
     * thing here, and why counting callbacks reads 60 on a box showing 15.
     */
    fun record(nowMs: Double, intervalMs: Double, presented: Boolean,
               cells: Int, width: Int, height: Int, track: String) {
        if (!visible) return
        ticks++
        if (!presented) { skipped++; return }
        if (intervalMs <= 0) return          // the first present has no interval
        // EVERY SERIES GETS EXACTLY ONE SAMPLE PER PRESENTED FRAME, so they stay
        // parallel to `stamps` and one trim is correct for all of them. A missing
        // value is recorded as 0 and filtered where it is read, never skipped —
        // skipping would slide one series against the others.
        stamps.add(nowMs)
        intervals.add(intervalMs)
        gpu.add(Ttp.ttp_display_gpu_ms())    // 0 = the query has not come back yet
        readProfile()
        trim(nowMs)
        if (nowMs - lastText < TEXT_INTERVAL_MS) return
        lastText = nowMs
        paint(cells, width, height, track)
        if (nowMs - lastLog >= LOG_INTERVAL_MS) {
            lastLog = nowMs
            // The ONLY channel a scripted sweep has. A TV has no console, and a
            // screenshot of this overlay cannot be diffed against another run.
            for (l in lines) Log.i("TtpPerf", l)
            logSpikes()
        }
    }

    /**
     * The renderer-CPU spike table, LOG-ONLY and at the log's own cadence — the
     * overlay keeps its median split, because a max painted at 4 Hz just
     * flickers. Two lines per log tick:
     *
     *  - `spike name:ms ...` — the full per-phase split of the window's WORST
     *    frame, worst by total+build: the same span the `ttp:render` atrace
     *    marker times, since buildFrame runs before render() and is outside
     *    `total`. This is the line that answers "which phase held the 8-20 ms
     *    frame", because the phases on it are from ONE frame and sum honestly.
     *  - `phasemax name:ms ...` — each phase's own max over the window, for the
     *    second-worst culprit and for spikes that never top a frame.
     *
     * `name:ms` with a COLON is the vocabulary `scripts/androidtv-live.mjs`
     * parses back out of the log — its per-phase spike attribution is keyed on
     * exactly this spelling.
     */
    private fun logSpikes() {
        val n = names ?: return
        val tot = sections["total"] ?: return
        if (tot.isEmpty()) return
        val build = sections["build"]
        var wi = 0
        var wv = -1.0
        for (i in tot.indices) {
            val v = tot[i] + (build?.getOrNull(i) ?: 0.0)
            if (v > wv) { wv = v; wi = i }
        }
        val cols = n.filter { !sections[it].isNullOrEmpty() }
        Log.i("TtpPerf", "spike " + cols.joinToString(" ") {
            "$it:${fmt(sections[it]?.getOrNull(wi) ?: 0.0)}" })
        Log.i("TtpPerf", "phasemax " + cols.joinToString(" ") {
            "$it:${fmt(sections[it]?.maxOrNull() ?: 0.0)}" })
    }

    /**
     * Drop what is older than the window, by count AND by age. Every series is cut
     * to the same length so a percentile over one is over the same frames as a
     * percentile over another.
     */
    private fun trim(nowMs: Double) {
        var drop = maxOf(0, stamps.size - WINDOW)
        while (drop < stamps.size && nowMs - stamps[drop] > WINDOW_MS) drop++
        if (drop == 0) return
        stamps.subList(0, drop).clear()
        intervals.subList(0, drop).clear()
        gpu.subList(0, drop).clear()
        for (v in sections.values) v.subList(0, minOf(drop, v.size)).clear()
    }

    private fun readProfile() {
        val buf = Ttp.ttp_display_profile()
        if (buf == null) {
            for (v in sections.values) v.add(0.0)   // keep the series parallel
            return
        }
        buf.order(ByteOrder.nativeOrder())
        val n = names ?: TtpJson.strOrEmpty(Ttp.ttp_display_profile_names())
            .split(",").also { names = it }
        for (i in n.indices) {
            if ((i + 1) * 8 > buf.capacity()) break
            // EVERY section, not just SHOWN: the spike table reads the whole
            // split of the window's worst frame, and the phase it needs is by
            // definition one the median display had no reason to show.
            sections.getOrPut(n[i]) { ArrayList(WINDOW) }.add(buf.getDouble(i * 8))
        }
    }

    /**
     * Budgets missed by a frame that took `interval`. Rounded rather than floored,
     * because presents land on vsyncs: a 25 ms interval is a frame that slipped
     * one budget, not 1.5 of them.
     */
    private fun missed(interval: Double): Int =
        if (interval > 0) maxOf(0, Math.round(interval / budgetMs).toInt() - 1) else 0

    private fun paint(cells: Int, width: Int, height: Int, track: String) {
        val recent = intervals.takeLast(60)
        val meanMs = if (recent.isEmpty()) 0.0 else recent.average()
        val fps = if (meanMs > 0) (1000.0 / meanMs) else 0.0
        val drops = recent.sumOf { missed(it) }
        val skipPct = if (ticks > 0) skipped * 100 / ticks else 0
        ticks = 0; skipped = 0
        val cpu = median(sections["total"]?.filter { it > 0 })
        val sorted = intervals.sorted()
        val p95 = if (sorted.isEmpty()) 0.0 else sorted[minOf(sorted.size - 1, (sorted.size * 0.95).toInt())]

        // The surface and the cadence on one line, the cost on the next, the split
        // under both. `cells` is here for the reason it is in the siblings: GPU
        // cost scales with cells and pixels together, so a logged number without
        // both is not comparable to any other logged number.
        val split = SHOWN.mapNotNull { s ->
            median(sections[s]?.filter { it > 0 })?.let { "$s ${fmt(it)}" } }
        lines = listOf(
            "${width}x$height · ${cells}c · ${"%.0f".format(Locale.ROOT, fps)} fps · " +
                "$drops drop${if (drops == 1) "" else "s"}${if (track.isEmpty()) "" else " · $track"}",
            "gpu ${share(median(gpu.filter { it > 0 }))} · cpu ${share(cpu)} · " +
                "present ${fmt(p95)}ms p95 · skip $skipPct%",
        ) + split.chunked(3).map { it.joinToString("  ") }

        // The web's thresholds, kept so the three readouts mean the same thing. With
        // no GPU timer the overshoot term is the interval's p95 past one budget.
        // DIAGNOSTIC colours: the sticker palette's veto on amber does not reach a
        // debug overlay, and matching the siblings is worth more than matching theme.
        val over = p95 / budgetMs - 1
        tint = when {
            drops > 2 || over > 1 || (fps > 0 && fps < 48) -> Color(0xFFFF6B6B)
            drops > 0 || over > 0.7 || (fps > 0 && fps < 57) -> Color(0xFFFFD166)
            else -> Color(0xFF7DFC8A)
        }
    }

    private fun fmt(ms: Double) = String.format(Locale.ROOT, "%.1f", ms)

    /**
     * A cost as milliseconds AND share of budget used, low is good — the same
     * sense as the two sibling readouts. The two costs do NOT sum: the CPU builds
     * frame N's commands while the GPU is still drawing N-1.
     */
    private fun share(ms: Double?) =
        ms?.let { "${fmt(it)}ms ${(it / budgetMs * 100).toInt()}%" } ?: "n/a"

    private fun median(xs: List<Double>?): Double? {
        if (xs.isNullOrEmpty()) return null
        return xs.sorted()[xs.size / 2]
    }
}

/**
 * Drop this over everything. It renders nothing at all while the monitor is
 * hidden, and it is never focusable — a focusable debug panel would steal remote
 * presses from the pause overlay and the results button.
 */
@Composable
fun PerfOverlay() {
    if (!PerfMonitor.visible) return
    Column(
        Modifier
            .padding(top = 24.dp, end = 24.dp)
            .background(Color(0xCC000000), RoundedCornerShape(8.dp))
            .padding(horizontal = 12.dp, vertical = 9.dp)
    ) {
        for (line in PerfMonitor.lines) {
            BasicText(
                line,
                style = TextStyle(
                    color = PerfMonitor.tint,
                    fontSize = 20.sp,
                    fontWeight = FontWeight.SemiBold,
                    fontFamily = FontFamily.Monospace,
                ),
            )
        }
    }
}
